### PR TITLE
Add remove flags for service update

### LINKED ISSUE #24064: service update of array/map types requires repeating the entire set
Issue for #23773

Updating array and mapping types by starting with an empty collection is fine when there are a small number of items, but it becomes a large burden for users if they have many ports, labels, or mounts, and would like to add or remove a single one. The case of adding or removing a single item should be a lot more common than removing everything.

We should design the UX around the more common case.

### PR DESCRIPTION
**Pull request by dnephin**
Following up on #23701

Updating array and mapping types by starting with an empty collection is fine when there are a small number of items, but it becomes a large burden for users if they have many ports, labels, or mounts, and would like to just add or remove a single one. The case of just adding or removing a single item should be a lot more common than removing everything.

This PR adds `--remove-x` flags and restores the old merge behaviour for these types.

Fixes #24064

cc @icecrime

### CODE REVIEW

**Comment by tonistiigi** [comment]
Does it make sense to replace the value if the mount target or exposed port matches?

**Comment by dnephin** [comment]
I think so, but I suspect that is done by the daemon somewhere already.

**Comment by aaronlehmann** [comment]
Design LGTM

File: api/client/service/update.go [hunk 1]
```diff
@@ -33,6 +34,11 @@ func newUpdateCommand(dockerCli *client.DockerCli) *cobra.Command {
 	flags.StringSlice("command", []string{}, "Service command")
 	flags.StringSlice("arg", []string{}, "Service command args")
 	addServiceFlags(cmd, opts)
+	flags.StringSlice(flagEnvRemove, []string{}, "Remove an environment variable")
+	flags.StringSlice(flagLabelRemove, []string{}, "The key of a label to remove")
```
Comment (aaronlehmann) [inline api/client/service/update.go]: I like the help text for removing environment variables above. "Remove an environment variable" puts the word "remove" front and center and uses an active voice that makes the description clearer. I'd suggest rephrasing this one as something like "Remove a label by its key", and so on for the others.

File: api/client/service/update.go [hunk 2]
```diff
@@ -157,6 +156,7 @@ func updateService(spec *swarm.ServiceSpec, flags *pflag.FlagSet) error {
 		updateDurationOpt((flagRestartWindow), task.RestartPolicy.Window)
 	}
 
+	// TODO: needs to be a merge
```
Comment (aaronlehmann) [inline api/client/service/update.go]: Let's handle this in this PR or in a followup before 1.12. We don't want to change the behavior later.

File: api/client/service/update.go [hunk 3]
```diff
@@ -189,65 +189,116 @@ func updateService(spec *swarm.ServiceSpec, flags *pflag.FlagSet) error {
...
+	if flags.Changed(flagEnv) {
+		value := flags.Lookup(flagEnv).Value.(*opts.ListOpts)
+		*field = append(*field, value.GetAll()...)
+	}
+	if flags.Changed(flagEnvRemove) {
+		toRemove, _ := flags.GetStringSlice(flagEnvRemove)
+		for _, envRemove := range toRemove {
+			for i, env := range *field {
+				key := envKey(env)
+				if key == envRemove {
```
Comment (aaronlehmann) [inline api/client/service/update.go]: It might be worth creating a utility function that converts a string slice to `map[string]struct{}`. Then these loop over `toRemove` that appear in each function can be simplified to map lookups.

File: api/client/service/update.go (hunk 1 above)
Comment (dnephin) [inline api/client/service/update.go]: That does sound better

File: api/client/service/update.go (hunk 3 above)
Comment (dnephin) [inline api/client/service/update.go]: sounds good

**Comment by tiborvass** [comment]
Sorry but i find this really confusing for users. I'd much rather have `update --label +foo=bar` and `update --label -foo`

**Comment by stevvooe** [comment]
> Sorry but i find this really confusing for users. I'd much rather have update --label +foo=bar and update --label -foo

Would it be sufficient to support `--label -foo=bar` to be the equivalent of `--label -foo`, or make the latter remove only the specified instance? For example, `--label -foo` would remove `foo=bar` or `foo=baz`, but `--label -foo=bar` would only remove `foo=bar`, but not `foo=baz`?

**Comment by dnephin** [comment]
I think the "accept key" or "accept full value" is a much smaller part of the issue. Supporting both would be fine with me.

I think the larger issue is to not wipe out the entire structure when you want to add or remove a single value from it, and how to do that with command line flags.

My feeling is that using `-` and `+` prefixes, which it seems simple, is not easy to discover. It also restricts the values we can use in the future because we now have these restrictions for updates.

Using `--remove-x` is obvious (you can see it in `--help`), and should be explicit (it's clear what you're try do).

**Comment by dnephin** [comment]
Changed the update/add flags to be `--x-add` and the remove flags to `--x-rm`

**Comment by aaronlehmann** [comment]
> My feeling is that using - and + prefixes, which it seems simple, is not easy to discover. It also restricts the values we can use in the future because we now have these restrictions for updates.

I agree.

> Changed the update/add flags to be --x-add and the remove flags to --x-rm

I'm happy with this solution.

Bikeshedding a little, but why this over `--add-x` / `--remove-x`? Putting the verb at the end feels backwards. I feel like I would type `--add-x` a lot and have to correct myself.

It probably doesn't matter much though.

**Comment by dnephin** [comment]
> Putting the verb at the end feels backwards.

The primary argument for this (based on some discussion) was that it makes `--help` display a lot better. The add and remove flags will be next to each other.

At first the prefix form felt more natural to me, but after implementing it this way, I think I'm happy with it as a suffix.

**Comment by dnephin** [comment]
This has been updated and is ready for another review

File: api/client/service/update.go [hunk 4]
```diff
@@ -189,65 +198,132 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
...
+	constraints := &placement.Constraints
+	placement.Constraints = append(placement.Constraints, field...)
 
-	localLabels := map[string]string{}
-	for key, value := range runconfigopts.ConvertKVStringsToMap(values) {
-		localLabels[key] = value
+	toRemove := buildToRemoveSet(flags, flagConstraintRemove)
+	for i, constraint := range placement.Constraints {
+		if _, exists := toRemove[constraint]; exists {
+			*constraints = append((*constraints)[:i], (*constraints)[i+1:]...)
```
Comment (aaronlehmann) [inline api/client/service/update.go]: Wondering if this is a safe operation. We're overwriting part of a slice with a shifted version of itself. Will `append` to the right thing? (In C terms, is it `memmove` or `memcpy`?)

File: api/client/service/update.go [hunk 5]
```diff
@@ -33,6 +34,19 @@ func newUpdateCommand(dockerCli *client.DockerCli) *cobra.Command {
...
 	flags.StringSlice("arg", []string{}, "Service command args")
 	addServiceFlags(cmd, opts)
+
+	flags.StringSlice(flagEnvRemove, []string{}, "Remove an environment variable")
+	flags.StringSlice(flagLabelRemove, []string{}, "Remove a label by its key")
+	flags.StringSlice(flagMountRemove, []string{}, "Remove a mount by its target path")
+	flags.StringSlice(flagPublishRemove, []string{}, "Remove a published port by its target port")
+	flags.StringSlice(flagNetworkRemove, []string{}, "Remove a network by name")
+	flags.StringSlice(flagConstraintRemove, []string{}, "Remove a constraint")
+	flags.VarP(&opts.labels, flagLabelAdd, "l", "Add or update service labels")
```
Comment (aaronlehmann) [inline api/client/service/update.go]: Not sure how I feel about the shorthand options for add being the same as `service create` but the long-form versions having the `-add` suffix. Any thoughts?

File: api/client/service/update.go (hunk 4 above)
Comment (dnephin) [inline api/client/service/update.go]: This is the "delete" recipe from https://github.com/golang/go/wiki/SliceTricks

I don't think it's overwriting part of the slice, it's assigning a new slice to the old variable from two parts of the original slice.

`slice.Delete(x)` would be great...

File: api/client/service/update.go (hunk 5 above)
Comment (dnephin) [inline api/client/service/update.go]: I wasn't sure either, if no one has a reason to keep them I can remove them.

File: api/client/service/update.go (hunk 4 above)
Comment (aaronlehmann) [inline api/client/service/update.go]: `append` will reuse space in the source slice if `len(x) < cap(x)` (which can be tricky in situations with concurrency, by the way).

Anyway, if the golang wiki says to do this, it sounds good to me.

File: api/client/service/update.go (hunk 5 above)
Comment (thaJeztah) [inline api/client/service/update.go]: if we're not sure about a short flag, I suggest to remove it, we can add later

Comment (thaJeztah) [inline api/client/service/update.go]: perhaps we should move adding the short flags to a separate PR, then we can discuss them separate from this one, and unlock this PR?

Comment (aaronlehmann) [inline api/client/service/update.go]: Don't let me slow down this PR. I was just raising the question.

Comment (thaJeztah) [inline api/client/service/update.go]: I think it's a valid question, so if we can split it from the main PR, that should be fine

**Comment by dnephin** [comment]
I've removed the short flags, and rebased

File: api/client/service/update.go [hunk 6]
```diff
@@ -33,6 +34,19 @@ func newUpdateCommand(dockerCli *client.DockerCli) *cobra.Command {
...
+	flags.StringSlice(flagMountRemove, []string{}, "Remove a mount by its target path")
+	flags.StringSlice(flagPublishRemove, []string{}, "Remove a published port by its target port")
+	flags.StringSlice(flagNetworkRemove, []string{}, "Remove a network by name")
+	flags.StringSlice(flagConstraintRemove, []string{}, "Remove a constraint")
+	flags.Var(&opts.labels, flagLabelAdd, "Add or update service labels")
+	flags.Var(&opts.env, flagEnvAdd, "Add or update environment variables")
+	flags.Var(&opts.mounts, flagMountAdd, "Add or update a mount on a service")
+	flags.StringSliceVar(&opts.constraints, flagConstraintAdd, []string{}, "Add or update placement constraints")
+	flags.StringSliceVar(&opts.networks, flagNetworkAdd, []string{}, "Add or update network attachments")
+	flags.Var(&opts.endpoint.ports, flagPublishAdd, "Add or update a publish port")
```
Comment (aaronlehmann) [inline api/client/service/update.go]: Should this be "published port"?

Comment (dnephin) [inline api/client/service/update.go]: sounds right, fixed

**Comment by aaronlehmann** [comment]
LGTM

**Comment by tiborvass** [comment]
Janky hitting #23401

File: api/client/service/update.go [hunk 7]
```diff
@@ -207,65 +216,132 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
...
-	for port := range ports {
-		localPortConfig = append(localPortConfig, convertPortToPortConfig(port, portBindings)...)
+	if flags.Changed(flagPublishRemove) {
+		toRemove, _ := flags.GetStringSlice(flagPublishRemove)
+		for _, rawTargetPort := range toRemove {
+			targetPort := nat.Port(rawTargetPort)
+			for i, port := range *portConfig {
+				if string(port.Protocol) == targetPort.Proto() &&
+					port.TargetPort == uint32(targetPort.Int()) {
+					*portConfig = append((*portConfig)[:i], (*portConfig)[i+1:]...)
```
Comment (cpuguy83) [inline api/client/service/update.go]: break?

Comment (dnephin) [inline api/client/service/update.go]: yes, I guess that would be more correct.

File: api/client/service/create.go [hunk 8]
```diff
@@ -27,7 +27,15 @@ func newCreateCommand(dockerCli *client.DockerCli) *cobra.Command {
 	flags := cmd.Flags()
 	flags.StringVar(&opts.mode, flagMode, "replicated", "Service mode (replicated or global)")
 	addServiceFlags(cmd, opts)
-	cmd.Flags().SetInterspersed(false)
+
+	flags.VarP(&opts.labels, flagLabel, "l", "Service labels")
+	flags.VarP(&opts.env, flagEnv, "e", "Set environment variables")
+	flags.VarP(&opts.mounts, flagMount, "m", "Attach a mount to the service")
```
Comment (tiborvass) [inline api/client/service/create.go]: @dnephin what is this `m`, is it a `-m` short flag ? If so, can we please not have it ?

Comment (dnephin) [inline api/client/service/create.go]: it is, sure I will remove it.

**Comment by tiborvass** [comment]
@dnephin needs a rebase, but overall looks good

**Comment by dongluochen** [comment]
This change doesn't allow multiple labels update within single `--label` flag or `--label-add` flag. But it allows removal of multiple labels with one `--label-rm` flag. Is this expected behavior? 

Ping @dnephin , @thaJeztah. 

```
dchen@vm2:~/docker-1.12$ docker service create --name test1 --label k1=v1,k2=v2 redis
11hma3k87jszuk9kv0b4lfqoo
dchen@vm2:~/docker-1.12$ docker service inspect test1 | grep -i label -A 2
            "Labels": {
                "k1": "v1,k2=v2"
            },
dchen@vm2:~/docker-1.12$ docker service update --label-add k3=v3 --label-add k4=v4 test1
test1
dchen@vm2:~/docker-1.12$ docker service inspect test1 | grep -i label -A 4
            "Labels": {
                "k1": "v1,k2=v2",
                "k3": "v3",
                "k4": "v4"
            },
dchen@vm2:~/docker-1.12$ docker service update --label-rm k1,k3 test1
test1
dchen@vm2:~/docker-1.12$ docker service inspect test1 | grep -i label -A 2
            "Labels": {
                "k4": "v4"
            },

dchen@vm2:~/docker-1.12$  docker version
Client:
 Version:      1.12.0-dev
 API version:  1.25
 Go version:   go1.6.2
 Git commit:   e7f699b
 Built:        Tue Jul 12 01:39:09 2016
 OS/Arch:      linux/amd64

dchen@vm2:~/go/src/github.com/docker/docker$ git log -n 3 --oneline
e7f699b Change the add/update flags to include 'add'
64ba8c6 PR feedback
364de29 Add remove flags for service update
```

**Comment by thaJeztah** [comment]
@dongluochen I agree, it's not consistent; left a comment here; https://github.com/docker/docker/pull/24397#issuecomment-232114842

ping @tiborvass for UX ^^

**Comment by thaJeztah** [comment]
temporarily moving back to design review, pending a decision on the comma-separated notation

**Comment by dongluochen** [comment]
The same Env parameter can be added multiple times. And it goes out of range at `env-rm`. 

```
dchen@vm2:~/go/src/github.com/docker/docker$ docker service create --name test1 --env PATH redis
6ciwaqa71m1reis5uhe3aeafw

dchen@vm2:~/go/src/github.com/docker/docker$ docker service inspect test1 | grep -i env -A 4
                    "Env": [
                        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/local/go/bin:/home/dchen/go/bin:/home/dchen/go/src/github.com/docker//swarm-v2/bin"
                    ]
                },

dchen@vm2:~/go/src/github.com/docker/docker$ docker service update --env-add SHELL --env-add PATH test1
test1

dchen@vm2:~/go/src/github.com/docker/docker$ docker service inspect test1 | grep -i env -A 5
                    "Env": [
                        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/local/go/bin:/home/dchen/go/bin:/home/dchen/go/src/github.com/docker//swarm-v2/bin",
                        "SHELL=/bin/bash",
                        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/local/go/bin:/home/dchen/go/bin:/home/dchen/go/src/github.com/docker//swarm-v2/bin"
                    ]
                },

dchen@vm2:~/go/src/github.com/docker/docker$ docker service update --env-rm PATH test1
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0xc37a00, 0xc820010090)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/docker/docker/api/client/service.updateEnvironment(0xc82007ea20, 0xc8200d2428)
	/go/src/github.com/docker/docker/api/client/service/update.go:270 +0x5bb
github.com/docker/docker/api/client/service.updateService(0xc82007ea20, 0xc8200d23c8, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/client/service/update.go:152 +0x30f
github.com/docker/docker/api/client/service.runUpdate(0xc820296b80, 0xc82007ea20, 0x7ffef5bb57bf, 0x5, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/client/service/update.go:63 +0x221
github.com/docker/docker/api/client/service.newUpdateCommand.func1(0xc820306000, 0xc820326b70, 0x1, 0x3, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/client/service/update.go:28 +0x6e
github.com/spf13/cobra.(*Command).execute(0xc820306000, 0xc820326ae0, 0x3, 0x3, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/spf13/cobra/command.go:593 +0x6fe
github.com/spf13/cobra.(*Command).ExecuteC(0xc8200818c0, 0xc820306000, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/spf13/cobra/command.go:683 +0x55c
github.com/spf13/cobra.(*Command).Execute(0xc8200818c0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/spf13/cobra/command.go:642 +0x2d
github.com/docker/docker/cli/cobraadaptor.CobraAdaptor.run(0xc8200818c0, 0xc820296b80, 0x7ffef5bb57a2, 0x7, 0xc82000a0e0, 0x4, 0x4, 0x0, 0x0)
	/go/src/github.com/docker/docker/cli/cobraadaptor/adaptor.go:119 +0x25f
github.com/docker/docker/cli/cobraadaptor.CobraAdaptor.Command.func1(0xc82000a0e0, 0x4, 0x4, 0x0, 0x0)
	/go/src/github.com/docker/docker/cli/cobraadaptor/adaptor.go:127 +0x8c
github.com/docker/docker/cli.(*Cli).Run(0xc820326570, 0xc82000a0d0, 0x5, 0x5, 0x0, 0x0)
	/go/src/github.com/docker/docker/cli/cli.go:81 +0x34b
main.main()
	/go/src/github.com/docker/docker/cmd/docker/docker.go:72 +0x599
```

**Comment by tiborvass** [comment]
For UX, let's not complicate and not implement `--label-rm k1,k2` and have only `--label-rm k1 --label-rm k2` for now. OK?

**Comment by dnephin** [comment]
The comma separated values are the default. Not accepting them makes it more complicated.

You can always do `--label-rm k1 --label-rm k2` . `StringSliceVar()` supports both by default.

**Comment by thaJeztah** [comment]
@dnephin but `--label` does *not* support comma-separated values;

```
docker run -dit --label foo=bar,bar=baz --name foobar alpine sh
docker inspect --format "{{json .Config.Labels }}" foobar

{"foo":"bar,bar=baz"}
```

**Comment by tiborvass** [comment]
@dnephin do we have something else than StringSliceVar? I really don't want `--label-rm k1,k2`.

**Comment by dnephin** [comment]
Sure, we can use `ListOpt`

**Comment by tiborvass** [comment]
@dnephin perfect, let's do that then :)

**Comment by dnephin** [comment]
Now using ListOpts for rm flags

> The same Env parameter can be added multiple times. And it goes out of range at

Thanks, fixed in the last commit.

**Comment by tiborvass** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐸

**Comment by vdemeester** [comment]
I'll make a follow-up for docs update (that way we don't run the CI again)
